### ISSUE TITLE
Counters not tracked in Redis when using floats

### ISSUE THREAD

**Issue by nussjustin-hmmh** [PromPHP/prometheus_client_php#39]
The Lua script for incrementing and tracking counters uses either `HINCRBY` or `HINCRBYFLOAT` depending on whether the value to increment by is an integer or a float.

The first time a counter is incremented the key is also added to a set of counters, which is then later used to find all counters.

To check whether a counter was newly created the Lua script compares the new counter value with the given value it was incremented by and if they are equal, considers the counter as new.

For this check the script converts the value to increment by to a number (using Luas `tonumber` function) but does not do the same for the result of the `HINCRBY` / `HINCRBYFLOAT` command.

In case of `HINCRYBY` the result will already be a number and the comparison will work, but `HINCRBYFLOAT` returns a string which causes the comparison to return false eventhough the numerical values are the same.

As a result the counter will not be added to the set of counters and can't be retrieved or rendered.

**Comment by evilworm** [comment]
I can confirm the issue, just stumbled upon this as well...
As a quick workaround you can compare if tostring(result) == tostring(ARGV[2]) in the updateCounter(...) method in Storage/Redis.php

**Comment by LKaemmerling** [comment]
@nussjustin-hmmh @evilworm thank you for the report, do you have time to submit an MR with a fix?

**Comment by evilworm** [comment]
tostring() unfortunately is not the answer since float precision in redis is 17 digits

```
incrbyfloat test 0.99999999999999999
"0.99999999999999999"
eval "return tostring(redis.call('get','test')) == tostring('0.9999999999999999999999999999999999')" 0
1
```

```
incrbyfloat test 0.9999999999999999999999999999999999
"1"
eval "return tostring(redis.call('get','test')) == tostring('0.9999999999999999999999999999999999')" 0
null
```

you would have to round the value to 17 digits, however I'm unable to get constant output from various float functions
```
eval "return string.format('%.17f', 0.99999)" 0
"0.99999000000000005"
incrbyfloat xxxxx 0.99999
"0.99999"
incrbyfloat yyyy 0.9999999999999999999999999999999999999999999999999999
"1"
eval "return string.format('%.17f', 0.99999999999999999999999999999999999999)" 0
"1.00000000000000000"
```

apparently incbyfloat does some kind of a hybrid form of float-string which is rounded with math.round if precision is greater than 17 digits... however internal string.format() function uses native form of floats and the precision "floats" as expected for the float type

not sure about the best solution here...

**Comment by nussjustin-hmmh** [comment]
I think the easiest solution would be running an `HEXISTS` before incrementing the value. The overhead should also be minimal.

**Comment by evilworm** [comment]
since it's a set, you should probably use `SISMEMBER` instead of `HEXISTS`

just tried this for updateCounter(), works great:
```
local result = redis.call(ARGV[1], KEYS[1], ARGV[3], ARGV[2])
if(redis.call('sIsMember', KEYS[2], KEYS[1]) == 0) then
    redis.call('hMSet', KEYS[1], '__meta', ARGV[4])
    redis.call('sAdd', KEYS[2], KEYS[1])
end
return result
```

I'm not sure about the performance penalty, but the documentation states that complexity for `SISMEMBER` is `O(1)` which means "constant time" ... so I think that's good enough, right?

**Comment by nussjustin-hmmh** [comment]
I was thinking of checking for the existence of the incremented float but checking the set also works. In that case instead ofusing `SISMEMBER` you could just call the `SADD` directly and check the return value (which will be the number of added values, so 1 means the value is new and 0 means the value was already in the set). The complexity is the same O(1) and should be as fast as `SISMEMBER` when the member is already in the set.

**Comment by EdgarSedov** [comment]
Just installed the library and wasted 2 hours to debug same issue) This bug existed for 15 months already, am i right?
![image](https://user-images.githubusercontent.com/16332716/109973648-bcc2e900-7d09-11eb-96ec-0637dadcc056.png)

**Comment by LKaemmerling** [comment]
@EdgarSedov @evilworm @nussjustin-hmmh  i just released `v2.2.2` which contains the fix from @nussjustin-hmmh for this issue.

**Comment by EdgarSedov** [comment]
@LKaemmerling Looks like now histograms are broken :)

Reproduce example:
```
$histogram = $registry->getOrRegisterHistogram(
    'namespace',
    'something',
    'description',
    [],
    [200, 500, 1000, 2000, 5000, 10000, 20000, 50000, 100000],
);

$histogram->observe(1003.5, []);
```

Counter with float works btw :)
